# Assignment 1

Assignment 1– Introduction to Python & sklearn 

Questions:
#1	[10 marks] MNIST (http://yann.lecun.com/exdb/mnist/) is a famous dataset and benchmark for pattern recognition. It contains images of hand written numbers that are normalized and centered in a 28x28 pixel array. The data set was derived from a NIST (National Institute of Standards and Technology) data set, hence the acronym for Modified NIST. There are many ways to load this data set. An easy way is to use routines from tensorflow to load 50 images into a Python tuple with 
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data', one_hot=True)
batch = mnist.train.next_batch(50)
Write a program that displays some of these examples.


#2	[20 marks] In the example code for the Iris data classification, we used 10-fold cross-validation to evaluate the accuracy of predictions with a linear SVM. In the example, we used the sklearn method model_selection.cross_val_score. 
a. Explain briefly what k-fold cross validation is and what it is used for. 
b. Write a script that does a k-fold cross-validation without using the cross_val_score function and compares the results with the sklearn function.   
c. Compare the cross-validated results of the SVM and RF and comment on which method is better.  


#3	[20 marks] Please download the wine.zip file and extract it to the directory for this assignment. Read through the wine_names.txt file and come to understand the problem and the wine data contained in the wine.train dataset. Train one of the models SVM, MLP, or RF to develop the best possible model for classifying the wine data in the hold-out test data set of 58 records in the wine.test file given the training data. In other words, you must submit a list of 58 classifications (as a separate *.csv file) for the hold-out test set in the same order as received. We will use your answers to score how well your model performs. 
Describe briefly your methodology for determining the best model and submit your final prediction program as well as the .csv file with the labels. Everyone will be ranked based on how well they do on their classification of the hold-out test set and 2 additional marks will be given to the best 10% of submissions.


## Question 1

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data', one_hot=True)

for i in range (0,5):
    batch = mnist.train.next_batch(1)
    print(batch[1])
    a=np.reshape(batch[0],(28,28))
    plt.imshow(a)
    plt.show()

C:\ProgramData\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


Extracting MNIST_data\train-images-idx3-ubyte.gz
Extracting MNIST_data\train-labels-idx1-ubyte.gz
Extracting MNIST_data\t10k-images-idx3-ubyte.gz
Extracting MNIST_data\t10k-labels-idx1-ubyte.gz
[[0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]]


[[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]]


[[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]]


[[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]]


[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


## Question 3

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import sklearn
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

wine_train = np.loadtxt('WINE_data\wine.train',delimiter=',')
wine_test = np.loadtxt('WINE_data\wine.test',delimiter=',')
#print(iris_data.shape)

XTrain = wine_train[:,1:]
YTrain = wine_train[:,0]

XTest = wine_test[:,1:]
YTest = wine_test[:,0]

# print(wine_test[:,])
# print(wine_test[0:,])

#svm KERNEL
model = svm.SVC(kernel='linear')
model.fit(XTrain,YTrain)
prediction=model.predict(XTest)
print("Prediction of Svm Kernel: ",prediction)

#RandomforestClassfier
clf = RandomForestClassifier()
clf.fit(XTrain,YTrain)
a=clf.predict(XTest)
print("Prediction of RandomForestClassifier: ",a)
'''np.savetxt('B00784526_prediction.csv',a,delimiter=',')'''

#MLPClassifier
from sklearn.neural_network import MLPClassifier
mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1)
mlp.fit(XTrain, YTrain)
MLPClassifier(activation='relu', alpha=1e-05, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False,
       epsilon=1e-08, hidden_layer_sizes=(5, 2), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=1, shuffle=True,
       solver='lbfgs', tol=0.0001, validation_fraction=0.1, verbose=False,
       warm_start=False)
prediction = mlp.predict(XTest)
print("Prediction of MLP: ", prediction)

Prediction of Svm Kernel:  [3. 1. 2. 2. 1. 2. 1. 2. 3. 2. 1. 1. 3. 3. 1. 1. 1. 2. 2. 1. 2. 1. 1. 2.
 1. 1. 2. 2. 1. 1. 1. 3. 1. 2. 3. 2. 2. 2. 2. 2. 2. 2. 1. 1. 2. 2. 1. 3.
 3. 2. 3. 2. 2. 2. 2. 1. 1. 2.]
Prediction of RandomForestClassifier:  [3. 1. 2. 2. 1. 2. 2. 2. 3. 2. 1. 1. 3. 3. 1. 1. 1. 2. 2. 2. 2. 1. 1. 2.
 1. 1. 2. 2. 1. 1. 1. 3. 1. 2. 3. 2. 2. 2. 2. 2. 2. 2. 1. 2. 2. 2. 1. 3.
 3. 2. 3. 2. 2. 2. 2. 1. 1. 2.]
Prediction of MLP:  [3. 1. 1. 2. 1. 2. 2. 2. 3. 2. 1. 1. 3. 2. 1. 1. 2. 2. 2. 2. 2. 1. 2. 2.
 1. 1. 1. 1. 1. 1. 1. 3. 2. 2. 3. 2. 1. 2. 2. 2. 2. 1. 1. 2. 2. 2. 1. 3.
 3. 2. 3. 2. 2. 2. 1. 2. 1. 3.]


SVC with linear kernel has been choosen to train the model for the following reasons:

-SVC works good on small data sets.
-overfitting is very low.
-Since there is very low overfitting, it can give us good predicitons on unseen data.

## Question 2

### Q2) [a]

a) K-fold cross validation: it is the way to find out how well out trained model would generalise on unseen and independent test data.
    In this the training dataset is divided into k equal parts or folds. Then each of the part becomes testing data atleast once, while the rest of the parts or folds are used as training set.
    Then the mean of the accuracy of the 'k' classifiers tells us how well our model would work on unseen data i.e. k-fold cross validation is used to find out the true accuracy of our model.

### Q2) [c]

In [3]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold, cross_val_score
from sklearn.ensemble import RandomForestClassifier

iris_data = np.loadtxt('IRIS_data\iris.data',delimiter=',')
print('iris_data shape: ',iris_data.shape)

    
train_x = iris_data[0::2,0:4]
train_y = np.array(iris_data[0::2, 4])
test_x = iris_data[1::2,0:4]
test_y = np.array(iris_data[0::2, 4])
kf = KFold(n_splits=10)

#SVM
clf = svm.SVC()
clf.fit(train_x,train_y)
print('\n\ntrainx: ',train_x.shape,len(train_x),'triany: ',train_y.shape,'testx: ',test_x.shape)
prediction=np.array(clf.predict(test_x))
print('prediction shape: ',prediction.shape)
print("accuracy: ", accuracy_score(test_y,prediction))
print('cross_val_score() gave the accuracy of: ',np.mean(cross_val_score(clf, train_x, train_y, cv=kf)))



#RF
clf = RandomForestClassifier()
clf.fit(train_x,train_y)
print('\n\ntrainx: ',train_x.shape,len(train_x),'triany: ',train_y.shape,'testx: ',test_x.shape)
prediction=np.array(clf.predict(test_x))
print('prediction shape: ',prediction.shape)
print("accuracy: ", accuracy_score(test_y,prediction))
print('RandomForestClassifier() gave the accuracy of: ',np.mean(cross_val_score(clf, train_x, train_y, cv=kf)))

iris_data shape:  (150, 5)


trainx:  (75, 4) 75 triany:  (75,) testx:  (75, 4)
prediction shape:  (75,)
accuracy:  0.9333333333333333
cross_val_score() gave the accuracy of:  0.9321428571428573


trainx:  (75, 4) 75 triany:  (75,) testx:  (75, 4)
prediction shape:  (75,)
accuracy:  0.96
RandomForestClassifier() gave the accuracy of:  0.9446428571428571


As we can see above RF gave the accuracy of 95.89% whereas SVM gave the accuarcy of 93.21%. RF works better on smaller datsets compared to SVM. The model is trained better using RF and the data is not overfitting unlike SVM.

### Q2) [b]

In [4]:
length=len(iris_data)
fold_size=int(length/10)

result=[]
finalresult =[]
iterator = iter(iris_data)
count=0
for i in range(length):
    if(count<14):
        result.append(iris_data[i])
        count=count+1
    else:
        count = 0
        result =[]
        finalresult.append(result)
folds = np.array(np.array_split(iris_data,10))

#diving the dataset into 10-fold structure.
CV_train_x = folds[:10,:,0:4]
CV_train_y = folds[:10,:,4:]

#shuffeling the dataset
import random
random.shuffle(CV_train_x)
random.shuffle(CV_train_y)

In [5]:
b=[]
predict=[]
accuracy=[]
for x in range(10):
    final_train_x = CV_train_x[np.arange(len(CV_train_x))!=x]
    final_train_y = CV_train_y[np.arange(len(CV_train_y))!=x]
    final_test_x = folds[x,:,0:4]
    final_test_y = folds[x,:,4:]
   
    #svm KERNEL
    model = svm.SVC()
    model.fit(final_train_x.reshape(135,4),np.ravel(final_train_y.reshape(135,1)))
    prediction=model.predict(final_test_x)
    predict.append(prediction)
    print("prediciton number",x," :",prediction)
    print("accuracy: ", accuracy_score(final_test_y,prediction))
    accuracy.append(accuracy_score(final_test_y,prediction))
 
# finding highest accuracy from our cross-val-score
list1 = accuracy
list1= set(accuracy)
list1 = list(list1)
length=len(list1)
list1.sort()
print("\n\n It gives the accuracy of: ", list1[length-2])

prediciton number 0  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 1  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 2  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 3  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 4  : [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
accuracy:  1.0
prediciton number 5  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 6  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 7  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 8  : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
accuracy:  1.0
prediciton number 9  : [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
accuracy:  0.6666666666666666


 It gives the accuracy of:  0.6666666666666666
